Mathlib/CategoryTheory/Join/Pseudofunctor.lean

```diff
@@ -0,0 +1,250 @@
+/-
+Copyright (c) 2025 Robin Carlier. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Robin Carlier
+-/
+import Mathlib.CategoryTheory.Join.Basic
+import Mathlib.CategoryTheory.Bicategory.Functor.Pseudofunctor
+
+/-!
+# Pseudofunctoriality of categorical joins
+
+In this file, we promote the join construction to two pseudofunctors
+`Join.pseudofunctorLeft` and `Join.pseudoFunctorRight`, expressing its pseudofunctoriality in
+each variable.
+
+-/
+
+universe v₁ v₂ u₁ u₂
+
+namespace CategoryTheory.Join
+
+open Bicategory
+
+/-- The `PrelaxFunctor` structure underlying `Join.pseudofunctorRight`. -/
+@[simps]
+def prelaxFunctorRight (C : Type u₁) [Category.{v₁} C] :
+    PrelaxFunctor Cat.{v₂, u₂} Cat.{max v₁ v₂, max u₁ u₂} where
+  obj D := Cat.of (C ⋆ D)
+  map F := mapPair (𝟭 C) F
+  map₂ := mapWhiskerLeft _
+  map₂_id {x y} f := by
+    apply natTrans_ext <;> aesop_cat
+  map₂_comp η θ := by
+    apply natTrans_ext <;>
+    ( dsimp
+      rw [← mapWhiskerLeft_comp] )
+
+/-- The `PrelaxFunctor` structure underlying `Join.pseudofunctorLeft`. -/
+@[simps]
+def prelaxFunctorLeft (D : Type u₂) [Category.{v₂} D] :
+    PrelaxFunctor Cat.{v₁, u₁} Cat.{max v₁ v₂, max u₁ u₂} where
+  obj C := Cat.of (C ⋆ D)
+  map F := mapPair F (𝟭 D)
+  map₂ := (mapWhiskerRight · _)
+  map₂_id {x y} f := by
+    apply natTrans_ext <;> aesop_cat
+  map₂_comp η θ := by
+    apply natTrans_ext <;>
+    ( dsimp
+      rw [← mapWhiskerRight_comp] )
+
+-- The proof gets too slow if we put it in a single `pseudofunctor` constructor,
+-- so we break down the component proofs for the pseudofunctors over several lemmas.
+
+/-- The structural isomorphism for composition of `pseudoFunctorRight`. -/
+def prelaxFunctorRight.mapCompRight (C : Type u₁) [Category.{v₁} C] {x y z : Cat.{v₂, u₂}}
```

Comment:
I would prefer if all the definitions were made in terms of functors between categories (possibly in different universe) rather than morphisms in `Cat`. If we have enough definitions and lemmas phrased in terms of `mapPair` (rather than `prelaxFunctorRight.map`, etc.), we would have a suitable API in heterogeneous universe situations, but at the same time, the definition of pseudofunctors would be easy.

Reply:
Yes. It seems that it also ease automation a lot.